File: Leetcode/0066-plus-one/solution.py
from typing import List

# ============================================================
# 0066. Plus One
# https://leetcode.com/problems/plus-one/
# Difficulty: Easy
# Tags: Array, Math
# ============================================================


class Solution:
    def plusOne(self, digits: List[int]) -> List[int]:
        """
        Optimal Solution (Walk + Carry).
        Time:  O(n)
        Space: O(1) amortized
        """
        carry = 1
        for i in range(len(digits) - 1, -1, -1):
            if carry == 0:
                break
            s = digits[i] + carry
            digits[i] = s % 10
            carry = s // 10
        if carry > 0:
            return [carry] + digits
        return digits

    def plusOneEarly(self, digits: List[int]) -> List[int]:
        n = len(digits)
        for i in range(n - 1, -1, -1):
            if digits[i] != 9:
                digits[i] += 1
                for j in range(i + 1, n): digits[j] = 0
                return digits
        return [1] + [0] * n


if __name__ == "__main__":
    sol = Solution()
    passed = failed = 0
    def test(name, got, expected):
        global passed, failed
        if got == expected: print(f"PASS: {name}"); passed += 1
        else: print(f"FAIL: {name} got={got}"); failed += 1

    cases = [
        ("Example 1", [1, 2, 3], [1, 2, 4]),
        ("Example 2", [4, 3, 2, 1], [4, 3, 2, 2]),
        ("Example 3", [9], [1, 0]),
        ("Single zero", [0], [1]),
        ("All nines 3", [9, 9, 9], [1, 0, 0, 0]),
        ("All nines 5", [9, 9, 9, 9, 9], [1, 0, 0, 0, 0, 0]),
        ("Trailing zeros", [1, 0, 0], [1, 0, 1]),
        ("Trailing nines", [1, 9, 9], [2, 0, 0]),
        ("Mid nines", [2, 9, 9, 1], [2, 9, 9, 2]),
        ("Long number", [1, 2, 3, 4, 5, 6, 7, 8, 9, 0], [1, 2, 3, 4, 5, 6, 7, 8, 9, 1]),
    ]

    print("=== Walk + Carry ===")
    for name, d, exp in cases:
        test(name, sol.plusOne(d.copy()), exp)

    print("\n=== Early Exit ===")
    for name, d, exp in cases:
        test("Early " + name, sol.plusOneEarly(d.copy()), exp)

    print(f"\nResults: {passed} passed, {failed} failed, {passed + failed} total")

File: Leetcode/0066-plus-one/solution.go
package main

import (
	"fmt"
	"reflect"
)

// ============================================================
// 0066. Plus One
// https://leetcode.com/problems/plus-one/
// Difficulty: Easy
// Tags: Array, Math
// ============================================================

// plusOne — Optimal Solution (Walk From End with Carry)
// Approach: scan right-to-left, propagate carry. Prepend 1 if carry remains.
// Time:  O(n)
// Space: O(1) amortized (O(n) only when prepending)
func plusOne(digits []int) []int {
	carry := 1
	for i := len(digits) - 1; i >= 0 && carry > 0; i-- {
		s := digits[i] + carry
		digits[i] = s % 10
		carry = s / 10
	}
	if carry > 0 {
		return append([]int{carry}, digits...)
	}
	return digits
}

// plusOneEarly — Find first non-9 from right, bump it, zero out the rest
// Time:  O(n)
// Space: O(1) / O(n)
func plusOneEarly(digits []int) []int {
	n := len(digits)
	for i := n - 1; i >= 0; i-- {
		if digits[i] != 9 {
			digits[i]++
			for j := i + 1; j < n; j++ {
				digits[j] = 0
			}
			return digits
		}
	}
	out := make([]int, n+1)
	out[0] = 1
	return out
}

// ============================================================
// Test Cases
// ============================================================

func cloneSlice(a []int) []int {
	b := make([]int, len(a))
	copy(b, a)
	return b
}

func main() {
	passed, failed := 0, 0
	test := func(name string, got, expected []int) {
		if reflect.DeepEqual(got, expected) {
			fmt.Printf("PASS: %s\n", name)
			passed++
		} else {
			fmt.Printf("FAIL: %s\n  Got:      %v\n  Expected: %v\n", name, got, expected)
			failed++
		}
	}

	type tc struct {
		name string
		in   []int
		want []int
	}
	cases := []tc{
		{"Example 1", []int{1, 2, 3}, []int{1, 2, 4}},
		{"Example 2", []int{4, 3, 2, 1}, []int{4, 3, 2, 2}},
		{"Example 3", []int{9}, []int{1, 0}},
		{"Single zero", []int{0}, []int{1}},
		{"All nines 3", []int{9, 9, 9}, []int{1, 0, 0, 0}},
		{"All nines 5", []int{9, 9, 9, 9, 9}, []int{1, 0, 0, 0, 0, 0}},
		{"Trailing zeros", []int{1, 0, 0}, []int{1, 0, 1}},
		{"Trailing nines", []int{1, 9, 9}, []int{2, 0, 0}},
		{"Mid nines", []int{2, 9, 9, 1}, []int{2, 9, 9, 2}},
		{"Long number", []int{1, 2, 3, 4, 5, 6, 7, 8, 9, 0}, []int{1, 2, 3, 4, 5, 6, 7, 8, 9, 1}},
	}

	fmt.Println("=== Walk + Carry ===")
	for _, c := range cases {
		test(c.name, plusOne(cloneSlice(c.in)), c.want)
	}
	fmt.Println("\n=== Early Exit ===")
	for _, c := range cases {
		test("Early "+c.name, plusOneEarly(cloneSlice(c.in)), c.want)
	}

	fmt.Printf("\nResults: %d passed, %d failed, %d total\n", passed, failed, passed+failed)
}
